docs/DesignDocs/AttestationApiRequirements.md

```diff
@@ -0,0 +1,317 @@
+
+Specific formats of Evidence, Attestation Results, etc. are the
+responsibility of the relevant plugins to implement, not the application.
+Similarly, plugins are responsible for TEE-specific details.
+
+The (non-attestation specific) protocol between the Attester and
+a Relying Party is the responsibility of the application, not
+the OE SDK or plugins.
+
+The (attestation-specific) protocol between a Verifier and either
+an Attester (in the Passport model) or a Relying Party (in the
+Background-check model) is the responsibility of the relevant
+plugins to implement, not the application.
+
+Specification
+-------------
+
+The APIs below are described as _abstract_ APIs, and any concrete
+APIs in C (or any other language) needs to accommodate an equivalent
+set of operations and support for an (at least) equivalent set of inputs
+and outputs.  This means the actual mapping of functions or arguments
+need not be 1:1 as long as a mapping exists.
+
+These APIs apply to both the "northbound" API exposed to apps by the SDK,
+as well as the "southbound" API between the SDK and plugins.
+
+# Attester APIs
+
+## GetEvidence call
+
+Inputs:
+
+* Requested Evidence format, if any (e.g., CWT, JWT, X.509, etc.).
+  Defaults to letting the plugin choose one.
+* Challenge buffer OCTET STRING
+* Include endorsements BOOLEAN
+* Custom claims buffer OCTET STRING
+
+Outputs:
+
+* Evidence buffer OCTET STRING
+* Format used for Evidence buffer
+
+Return status codes:
+
+* Success indicates that Evidence was generated as requested
+* Failed-to-get-endorsements indicates that the call failed because
+  "Include endorsements" was true but that endorsements could not be
+  obtained
+* Requested-format-not-supported indicates that the requested Evidence
+  format could not be used
+* Challenge-Parse-error indicates that the challenge buffer provided could not
+  be correctly parsed
+* Custom-Claims-Parse-error indicates that the custom claims buffer provided
+  could not be correctly parsed per the requested Evidence format
+* Other failure indicates that the call failed for any other reason
+
+This is the main attestation functionality of the Attester.
+The challenge buffer, which may be empty for some protocols (e.g.,
+those that use timestamps and synchronized clocks to check freshness),
+can be provided by the app if the app has its own way to get it
+from a Verifier.  For example, in the example in Section 15.4 of
+[RATS Architecture](https://tools.ietf.org/wg/rats/draft-ietf-rats-architecture/),
+it is obtained via the (non-attestation specific) protocol between
+the Attester and the Relying Party.  (This buffer is first created
+by a Verifier in its GetChallenge call discussed below, but might be
+communicated to an Attester in various ways.)
+
+Note that Section 15.2 contains a Nonce-based Passport model example
+where the challenge is obtained across the Attester-Verifier protocol,
+but that protocol is the responsibility of the plugin, not the app
```

Comment:
Is the verifier plugin then responsible for exposing the API to get nonce which the attester application can turn into a custom claim when calling GetEvidence()?

Reply:
GetChallenge (line 189) represents that API for verifier applications.   Of course the verifier is not required to use the OE SDK per se, but if it does then yes that is the intended abstract API.